include/rocksdb/customizable.h

```diff
@@ -0,0 +1,149 @@
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+// Copyright (c) 2011 The LevelDB Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file. See the AUTHORS file for names of contributors.
+
+#pragma once
+
+#include "rocksdb/configurable.h"
+#include "rocksdb/status.h"
+
+namespace ROCKSDB_NAMESPACE {
+/**
+ * Customizable a base class used by the rocksdb that describes a
+ * standard way of configuring and creating objects.  Customizable objects
+ * are configurable objects that can be created from an ObjectRegistry.
+ *
+ * When a Customizable is being created, the "name" property specifies
+ * the name of the instance being created.
+ * For custom objects, their configuration and name can be specified by:
+ * [prop]={name=X;option 1 = value1[; option2=value2...]}
+ *
+ * [prop].name=X
+ * [prop].option1 = value1
+ *
+ * [prop].name=X
+ * X.option1 =value1
+ */
+class Customizable : public Configurable {
+ public:
+  virtual ~Customizable() {}
+  // Returns the name of this class of Customizable
+  virtual const char* Name() const = 0;
+  // Returns an identifier for this Customizable.
+  // This could be its name or something more complex (like its URL/pattern).
+  // Used for pretty printing.
+  virtual std::string GetId() const {
+    std::string id = Name();
+    return id;
+  }
+
+  // This is typically determined by if the input name matches the
+  // name of this object.
+  // This method is typically used in conjunction with CastAs to find the
+  // derived class instance from its base.  For example, if you have an Env
+  // and want the "Default" env, you would IsInstanceOf("Default") to get
+  // the default implementation.  This method should be used when you need a
+  // specific derivative or implementation of a class.
+  //
+  // Intermediary caches (such as SharedCache) may wish to override this method
+  // to check for the intermediary name (SharedCache).  Classes with multiple
+  // potential names (e.g. "PosixEnv", "DefaultEnv") may also wish to override
+  // this method.
+  //
+  // @param name The name of the instance to find.
+  // Returns true if the class is an instance of the input name.
+  virtual bool IsInstanceOf(const std::string& name) const {
+    return name == Name();
+  }
+
+  // Returns the named instance of the Customizable as a T*, or nullptr if not
+  // found. This method uses IsInstanceOf to find the appropriate class instance
+  // and then casts it to the expected return type.
+  template <typename T>
+  const T* CastAs(const std::string& name) const {
+    if (IsInstanceOf(name)) {
+      return static_cast<const T*>(this);
+    } else {
+      const auto inner = static_cast<const Customizable*>(Inner());
+      if (inner != nullptr) {
```

Comment:
I'm a little bit confused here. If the name does not match, why we need to continue the inner configurable and try if the name matches the instance.

Reply:
There are Customizable classes that have a has-a relationship.  For example, an EncryptedEnv has-a PosixEnv in it (as the inner).  If I needed access to the underlying posix file system, there is no way to get to it without going through the inner class.  Or a DBWithTTL has a DBImpl in it somewhere and you need a means of getting at the underlying implementation.

Perhaps it might be better to rename the "CastAs" method to something like "GetAs"?  Would that help alleviate some of the confusion?

Reply:
Maybe `GetAs` is better. CastAs is something like we force to change its type to.